### ISSUE TITLE
Best Practice for Creating Arabic Dictionary

### ISSUE THREAD

**Issue by Hegelim** [PaddlePaddle/PaddleOCR#10431]
## Question
Let's say I want to train an Arabic recognition model, what's the best practice when creating a customized Arabic dictionary?
Now, there are several things that make it challenging:
1. Arabic letters change their shapes depending on their locations in the word, for example, the letter alif has 4 forms and each one has a unicode glyph. Should I include all possible shapes of it in the dictionary or should I just include a single letter in the alphabet?
2. follow-up on 1, if I only include a single letter, then how is the model trained such that it can recognize different shapes of the same letter? It sounds like a 1-to-many mapping, can the model do that?
3. Arabic is cursive, that means when joining letters together, they merge together, which is called ligature. How can I take this into account when creating the dictionary?
4. What's the order of paddleOCR recognition? Because Arabic is a right-to-left language, and if paddleocr reads texts from left to right, should I be concerned and are there any files that I should change? 

## 系统环境/System Environment：
Ubuntu 18.04

## 版本号/Version：Paddle：  PaddleOCR： 问题相关组件/Related components：
paddleocr 2.6.1.0
paddlepaddle-gpu 2.4.2

**Comment by shiyutang** [comment]
"When creating a customized Arabic dictionary for training an Arabic recognition model, it is recommended to consider the following best practices:

1. Including Multiple Shapes of Arabic Letters: Since Arabic letters can change their shapes depending on their positions in a word, it is advisable to include all possible shapes of each letter in the dictionary. This means including the different forms of letters like alif, such as initial, medial, final, and isolated forms. Each form should have its corresponding Unicode glyph representation.

2. Training the Model for Recognizing Different Shapes: To train the model to recognize different shapes of the same letter, you can use data augmentation. By generating synthetic images with variations in letter shapes, you can provide the model with diverse examples to learn from. Additionally, you can use techniques like elastic distortion or geometric transformations to further enhance the model's ability to recognize different letter shapes.

3. Handling Ligatures and Cursive Writing: Arabic script often involves ligatures, where letters merge when joined in a word. When creating the dictionary, it is crucial to include ligatures as separate entries. You can represent ligatures as individual characters or use a special symbol to denote ligature formation. This allows the model to learn and recognize ligatures during training and inference.

4. Order of Recognition in PaddleOCR: PaddleOCR follows a left-to-right recognition order by default. However, for Arabic text, which is read from right to left, you may need to modify certain files to ensure correct recognition. You should check and potentially modify the order of character processing within the model architecture, as well as any post-processing steps that reorder the recognized text. By adapting the recognition pipeline to handle right-to-left languages, you can ensure accurate Arabic text recognition with PaddleOCR.

In summary, include all possible shapes of Arabic letters in the dictionary, provide diverse training data for different letter forms, include ligature forms, and rely on PaddleOCR's built-in support for right-to-left languages."

**Comment by ruogedol** [comment]
If you need to identify the normal results, you can refer to **https://github.com/mpcabd/python-arabic-reshaper** and install the **arabic_reshaper** package in the Python environment. 
Copy the recognition results to a text editor, and the editor's plugin may automatically correct the word order.
The issues shown do not affect model training.

In version 2.5, because of the special variation of Arabic characters, I used a single font to generate word data to train, and the perfect pair rate of recognizing words on PC was 95%. But with this best result, the model is not good to transfer training long sentences.

**Comment by roostaeecs** [comment]
I am using PaddleOCR with the Arabic language model (lang='fa') to perform OCR on Arabic images. While PaddleOCR correctly recognizes the Arabic characters, it processes the text in a Left-to-Right (LTR) order, which is incorrect for Arabic, a Right-to-Left (RTL) language. This results in the words and sentences being in reverse order.

I have reviewed the paddleocr --help output to see if there are any options to explicitly set the text direction or handle RTL languages like Arabic.

My question is:

Is there a specific option in PaddleOCR, possibly using ocr_order_method or another parameter, to correctly handle Right-to-Left languages like Arabic and ensure the output text is in the correct RTL order?

If there isn't a built-in option, what are the recommended workarounds to post-process the OCR output to reorder the text correctly for RTL languages in Python?

Any guidance or solutions on how to get PaddleOCR to output Arabic text in the correct Right-to-Left order would be greatly appreciated.

I tried to use the following code:

from paddleocr import PaddleOCR, draw_ocr

ocr = PaddleOCR(use_angle_cls=True, lang='Ar') 
img_path = 'image5.jpg'
result = ocr.ocr(img_path, cls=True)
for idx in range(len(result)):
    res = result[idx]
    for line in res:
        print(line)

# draw result
from PIL import Image
result = result[0]
image = Image.open(img_path).convert('RGB')
boxes = [line[0] for line in result]
txts = [line[1][0] for line in result]
scores = [line[1][1] for line in result]
im_show = draw_ocr(image, boxes, txts, scores, font_path='./doc/fonts/arabic.ttf')
im_show = Image.fromarray(im_show)
im_show.save('result.jpg')
